### PR TITLE
[TagInput] onAdd can return false to prevent clearing input

### LINKED ISSUE #1277: [Labs] Only unique values in TagInput
Possibility to have only unique values in TagInput component.

Concerns:
- Should we support this out of the box? Maybe it's responsible of user code which controls values passed to `TagInput` component?
- Should we call some kind of method from props, when I'm trying to add value that already exists?
- Should we define uniqueness only base on value?

### PR DESCRIPTION
**Pull request by cmslewis**
#### Checklist

- [x] Include tests

#### Changes proposed in this pull request:

- **Changed** `TagInput` `onAdd` now returns a `boolean`, that, if `false`, will cause the input NOT to be cleared.
    - This is useful, for example, if the provided `value` is somehow invalid and should not be added as a tag (though whether or not the tag is added is the consumer's concern, due to the strictly controlled nature of `TagInput`).

#### Reviewers should focus on:

Does this behavior seem reasonable?

### CODE REVIEW

File: packages/labs/src/tag-input/tagInput.tsx [hunk 1]
```diff
@@ -167,8 +170,10 @@ export class TagInput extends AbstractComponent<ITagInputProps, ITagInputState>
         const { selectionEnd, value } = event.currentTarget;
         if (event.which === Keys.ENTER && value.length > 0) {
             // enter key on non-empty string invokes onAdd
-            this.setState({ inputValue: "" });
-            Utils.safeInvoke(this.props.onAdd, value);
+            const shouldClearInput = Utils.safeInvoke(this.props.onAdd, value);
+            if (shouldClearInput) {
```
Comment (llorca) [inline packages/labs/src/tag-input/tagInput.tsx]: Let's add a switch in the examples for "Unique values" 👍 

Can also come as a follow up PR, happy to take care of this

File: packages/labs/src/tag-input/tagInput.tsx [hunk 2]
```diff
@@ -19,8 +19,11 @@ export interface ITagInputProps extends IProps {
      * Callback invoked when a new tag is added by the user pressing `enter` on the input.
      * Receives the current value of the input field. New tags are expected to be appended to
      * the list.
+     *
+     * If the provided function returns `false`, the input will not be cleared. This is useful,
+     * for example, if the provided `value` is somehow invalid and should not be added as a tag.
      */
-    onAdd?: (value: string) => void;
+    onAdd?: (value: string) => boolean;
```
Comment (llorca) [inline packages/labs/src/tag-input/tagInput.tsx]: What about `boolean | void`, so that you can return both `true` or nothing if you want the tag to be added?

Comment (cmslewis) [inline packages/labs/src/tag-input/tagInput.tsx]: Ah yep, missed that. Fixed. 👍

File: packages/labs/src/tag-input/tagInput.tsx (hunk 1 above)
Comment (cmslewis) [inline packages/labs/src/tag-input/tagInput.tsx]: I'm iffy on adding a "unique values" switch, because it's doesn't correspond out-of-the-box behavior. Would love to figure out a good way to communicate how to do that in a future PR though. Maybe we could just include a code snippet similar to `NumericInput`s section on uncontrolled usage:

http://blueprintjs.com/docs/#core/components/forms/numeric-input.uncontrolled-mode

Comment (llorca) [inline packages/labs/src/tag-input/tagInput.tsx]: That's fair, I mean I guess the docs speaks for itself now, we really were lacking this `return false` capability

File: packages/labs/src/tag-input/tagInput.tsx [hunk 3]
```diff
@@ -167,8 +171,13 @@ export class TagInput extends AbstractComponent<ITagInputProps, ITagInputState>
         const { selectionEnd, value } = event.currentTarget;
         if (event.which === Keys.ENTER && value.length > 0) {
             // enter key on non-empty string invokes onAdd
-            this.setState({ inputValue: "" });
-            Utils.safeInvoke(this.props.onAdd, value);
+            const shouldClearInput = Utils.safeInvoke(this.props.onAdd, value);
+
+            // strict equal is necessary to ensure the input is cleared whether the return value is
+            // true or undefined
```
Comment (giladgray) [inline packages/labs/src/tag-input/tagInput.tsx]: PR title does not match implementation. "Return boolean from onAdd that [does not clear input if false]"

File: packages/labs/src/tag-input/tagInput.tsx [hunk 4]
```diff
@@ -19,8 +19,12 @@ export interface ITagInputProps extends IProps {
      * Callback invoked when a new tag is added by the user pressing `enter` on the input.
      * Receives the current value of the input field. New tags are expected to be appended to
      * the list.
+     *
+     * By default, the input will be cleared after `onAdd` is invoked. However, if the `onAdd`
+     * function returns `false`, the input will _not_ be cleared. This is useful, for example,
+     * if the provided `value` is somehow invalid and should not be added as a tag.
```
Comment (giladgray) [inline packages/labs/src/tag-input/tagInput.tsx]: shorter, same idea:

> The input will be cleared after `onAdd` is invoked _unless_ the callback explicitly returns `false`. This is useful if the provided `value` is somehow invalid and should not be added as a tag.

File: packages/labs/src/tag-input/tagInput.tsx [hunk 5]
```diff
@@ -167,8 +171,13 @@ export class TagInput extends AbstractComponent<ITagInputProps, ITagInputState>
         const { selectionEnd, value } = event.currentTarget;
         if (event.which === Keys.ENTER && value.length > 0) {
             // enter key on non-empty string invokes onAdd
-            this.setState({ inputValue: "" });
-            Utils.safeInvoke(this.props.onAdd, value);
+            const shouldClearInput = Utils.safeInvoke(this.props.onAdd, value);
+
```
Comment (giladgray) [inline packages/labs/src/tag-input/tagInput.tsx]: unnecessary whitespace as next block is directly related to variable definition: continuing the thought.

File: packages/labs/src/tag-input/tagInput.tsx (hunk 3 above)
Comment (giladgray) [inline packages/labs/src/tag-input/tagInput.tsx]: how about pithy "only explicit return false cancels text clearing"

**Comment by cmslewis** [comment]
@giladgray ready for re-review
